root/roofit/vectorisedPDFs/benchAddPdf.cxx

```diff
@@ -92,15 +100,19 @@ static void benchAddPdfGaussExp(benchmark::State& state) {
   RooBatchCompute::RunContext evalData;
   std::vector<double> results(nEvents);
 
+  std::array<RooArgSet, nParamSets> paramSets;
+  unsigned int seed = 1337;
+  for (auto& paramSet : paramSets) {
+    randomiseParameters(parameters, seed++);
+    parameters.snapshot(paramSet);
+  }
+
   for (auto _ : state) {
-    for (unsigned int paramSetIndex=0; paramSetIndex < nParamSets; ++paramSetIndex) {
-      state.PauseTiming();
-      randomiseParameters(parameters, 1337+paramSetIndex);
-      state.ResumeTiming();
+    for (const auto& paramSet : paramSets) {
+      parameters = paramSet;
```

Comment:
Nice idea! I guess that's the way to avoid that any intermediate results are cached and that fits are measured to be very fast because the parameter values are already in the minimum for the later benchmark iterations.

Reply:
There's actually no fit here; it's just evaluating. But the parameters are changed to make sure that no caches are used and that the benchmark goes via different code paths (in case the PDF uses `if (val > x) ...; else ...`).

Reply:
Ah yes, I see. I was thinking also about fitting because I think I need to implement a randomization like this also in the other benchmarks. Here, it calls the minimizer in a loop for example: https://github.com/root-project/rootbench/blob/master/root/roofit/roofit/RooFitBinnedBenchmarks.cxx#L137 As far as I understand, starting the minimization from the result of the previous minimization is not a good benchmark.

Reply:
Correct. But there's three things:
1. You want to bench evaluation. In this case, just kick all parameters a bit.
2. You want to bench evaluation & caching: Kick only one parameter (as the minimiser would do during gradient calculation), then kick all (line search).
3. You went to bench evaluation and minimiser: Use a minimiser. Here it's presumably OK to kick one or two parameters a bit before starting the minimisation.